drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -0,0 +1,1680 @@
+{
+	u32_t sys_cfg;
+	u16_t sys_ctrl = DWT_SYS_CTRL_RXENAB;
+
+	sys_cfg = dwt_reg_read_u32(ctx, DWT_SYS_CFG_ID, 0);
+
+	if (timeout != 0) {
+		dwt_reg_write_u16(ctx, DWT_RX_FWTO_ID, DWT_RX_FWTO_OFFSET,
+				  timeout);
+		sys_cfg |= DWT_SYS_CFG_RXWTOE;
+	} else {
+		sys_cfg &= ~DWT_SYS_CFG_RXWTOE;
+	}
+
+	dwt_reg_write_u32(ctx, DWT_SYS_CFG_ID, 0, sys_cfg);
+	dwt_reg_write_u16(ctx, DWT_SYS_CTRL_ID, DWT_SYS_CTRL_OFFSET, sys_ctrl);
+
+	return 0;
+}
+
+static inline void dwt_irq_handle_rx_cca(struct dwt_context *ctx)
+{
+	k_sem_give(&ctx->phy_sem);
+	ctx->cca_busy = true;
+
+	/* Clear all RX event bits */
+	dwt_reg_write_u32(ctx, DWT_SYS_STATUS_ID, 0,
+			  DWT_SYS_STATUS_ALL_RX_GOOD);
+}
+
+static inline void dwt_irq_handle_rx(struct dwt_context *ctx, u32_t sys_stat)
+{
+	struct net_pkt *pkt = NULL;
+	struct dwt_rx_info_regs rx_inf_reg;
+	float a_const;
+	u32_t rx_finfo;
+	u32_t ttcki;
+	u32_t rx_pacc;
+	u32_t cir_pwr;
+	u32_t flags_to_clear;
+	s32_t ttcko;
+	u16_t pkt_len;
+	u8_t *fctrl;
+	s8_t rx_level = INT8_MIN;
+
+	LOG_DBG("RX OK event, SYS_STATUS 0x%08x", sys_stat);
+	flags_to_clear = sys_stat & DWT_SYS_STATUS_ALL_RX_GOOD;
+
+	rx_finfo = dwt_reg_read_u32(ctx, DWT_RX_FINFO_ID, DWT_RX_FINFO_OFFSET);
+	pkt_len = rx_finfo & DWT_RX_FINFO_RXFLEN_MASK;
+	rx_pacc = (rx_finfo & DWT_RX_FINFO_RXPACC_MASK) >>
+		   DWT_RX_FINFO_RXPACC_SHIFT;
+
+	if (!(IS_ENABLED(CONFIG_IEEE802154_RAW_MODE))) {
+		pkt_len -= DWT_FCS_LENGTH;
+	}
+
+	pkt = net_pkt_alloc_with_buffer(ctx->iface, pkt_len,
+					AF_UNSPEC, 0, K_NO_WAIT);
+	if (!pkt) {
+		LOG_ERR("No buf available");
+		goto rx_out_enable_rx;
+	}
+
+	dwt_register_read(ctx, DWT_RX_BUFFER_ID, 0, pkt_len, pkt->buffer->data);
+	dwt_register_read(ctx, DWT_RX_FQUAL_ID, 0, sizeof(rx_inf_reg),
+			  (u8_t *)&rx_inf_reg);
+	net_buf_add(pkt->buffer, pkt_len);
+	fctrl = pkt->buffer->data;
+
+	/*
+	 * Get Ranging tracking offset and tracking interval
+	 * for Crystal characterization
+	 */
+	ttcki = sys_get_le32(rx_inf_reg.rx_ttcki);
+	ttcko = sys_get_le32(rx_inf_reg.rx_ttcko) & DWT_RX_TTCKO_RXTOFS_MASK;
+	/* Traking offset value is a 19-bit signed integer */
+	if (ttcko & BIT(18)) {
+		ttcko |= ~DWT_RX_TTCKO_RXTOFS_MASK;
+	}
+
+	/* TODO add:
+	 * net_pkt_set_ieee802154_tcki(pkt, ttcki);
+	 * net_pkt_set_ieee802154_tcko(pkt, ttcko);
+	 */
+	LOG_DBG("ttcko %d ttcki: 0x%08x", ttcko, ttcki);
+
+	if (IS_ENABLED(CONFIG_NET_PKT_TIMESTAMP)) {
+		u8_t ts_buf[sizeof(u64_t)] = {0};
+		struct net_ptp_time timestamp;
+		u64_t ts_fsec;
+
+		memcpy(ts_buf, rx_inf_reg.rx_time, DWT_RX_TIME_RX_STAMP_LEN);
+		ts_fsec = sys_get_le64(ts_buf) * DWT_TS_TIME_UNITS_FS;
+		timestamp.second = (ts_fsec / 1000000) / NSEC_PER_SEC;
+		timestamp.nanosecond = (ts_fsec / 1000000) % NSEC_PER_SEC;
+		net_pkt_set_timestamp(pkt, &timestamp);
+	}
+
+	/* See 4.7.2 Estimating the receive signal power */
+	cir_pwr = sys_get_le16(&rx_inf_reg.rx_fqual[6]);
+	if (ctx->rf_cfg.prf == DWT_PRF_16M) {
+		a_const = 113.77;
+	} else {
+		a_const = 121.74;
```

Comment:
magic numbers, could you move them to macros which would give some type of hint on what it is exactly?

Reply:
fixed